## Problem 8: Largest Product in a Series

The four adjacent digits in the 1000-digit number that have the greatest product are $9 \times 9 \times 8 \times 9 = 5832$.

![](./images/image1.png){width=65%}


Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?

Let's solve using **Julia**

We can start by putting all of the digits into a vector which we'll call `all_digits`

In [1]:
all_digits = [7,3,1,6,7,1,7,6,5,3,1,3,3,0,6,2,4,9,1,9,2,2,5,1,1,9,6,7,4,4,2,6,5,7,4,7,4,2,3,5,5,3,4,9,1,9,4,9,3,4,
              9,6,9,8,3,5,2,0,3,1,2,7,7,4,5,0,6,3,2,6,2,3,9,5,7,8,3,1,8,0,1,6,9,8,4,8,0,1,8,6,9,4,7,8,8,5,1,8,4,3,
              8,5,8,6,1,5,6,0,7,8,9,1,1,2,9,4,9,4,9,5,4,5,9,5,0,1,7,3,7,9,5,8,3,3,1,9,5,2,8,5,3,2,0,8,8,0,5,5,1,1,
              1,2,5,4,0,6,9,8,7,4,7,1,5,8,5,2,3,8,6,3,0,5,0,7,1,5,6,9,3,2,9,0,9,6,3,2,9,5,2,2,7,4,4,3,0,4,3,5,5,7,
              6,6,8,9,6,6,4,8,9,5,0,4,4,5,2,4,4,5,2,3,1,6,1,7,3,1,8,5,6,4,0,3,0,9,8,7,1,1,1,2,1,7,2,2,3,8,3,1,1,3,
              6,2,2,2,9,8,9,3,4,2,3,3,8,0,3,0,8,1,3,5,3,3,6,2,7,6,6,1,4,2,8,2,8,0,6,4,4,4,4,8,6,6,4,5,2,3,8,7,4,9,
              3,0,3,5,8,9,0,7,2,9,6,2,9,0,4,9,1,5,6,0,4,4,0,7,7,2,3,9,0,7,1,3,8,1,0,5,1,5,8,5,9,3,0,7,9,6,0,8,6,6,
              7,0,1,7,2,4,2,7,1,2,1,8,8,3,9,9,8,7,9,7,9,0,8,7,9,2,2,7,4,9,2,1,9,0,1,6,9,9,7,2,0,8,8,8,0,9,3,7,7,6,
              6,5,7,2,7,3,3,3,0,0,1,0,5,3,3,6,7,8,8,1,2,2,0,2,3,5,4,2,1,8,0,9,7,5,1,2,5,4,5,4,0,5,9,4,7,5,2,2,4,3,
              5,2,5,8,4,9,0,7,7,1,1,6,7,0,5,5,6,0,1,3,6,0,4,8,3,9,5,8,6,4,4,6,7,0,6,3,2,4,4,1,5,7,2,2,1,5,5,3,9,7,
              5,3,6,9,7,8,1,7,9,7,7,8,4,6,1,7,4,0,6,4,9,5,5,1,4,9,2,9,0,8,6,2,5,6,9,3,2,1,9,7,8,4,6,8,6,2,2,4,8,2,
              8,3,9,7,2,2,4,1,3,7,5,6,5,7,0,5,6,0,5,7,4,9,0,2,6,1,4,0,7,9,7,2,9,6,8,6,5,2,4,1,4,5,3,5,1,0,0,4,7,4,
              8,2,1,6,6,3,7,0,4,8,4,4,0,3,1,9,9,8,9,0,0,0,8,8,9,5,2,4,3,4,5,0,6,5,8,5,4,1,2,2,7,5,8,8,6,6,6,8,8,1,
              1,6,4,2,7,1,7,1,4,7,9,9,2,4,4,4,2,9,2,8,2,3,0,8,6,3,4,6,5,6,7,4,8,1,3,9,1,9,1,2,3,1,6,2,8,2,4,5,8,6,
              1,7,8,6,6,4,5,8,3,5,9,1,2,4,5,6,6,5,2,9,4,7,6,5,4,5,6,8,2,8,4,8,9,1,2,8,8,3,1,4,2,6,0,7,6,9,0,0,4,2,
              2,4,2,1,9,0,2,2,6,7,1,0,5,5,6,2,6,3,2,1,1,1,1,1,0,9,3,7,0,5,4,4,2,1,7,5,0,6,9,4,1,6,5,8,9,6,0,4,0,8,
              0,7,1,9,8,4,0,3,8,5,0,9,6,2,4,5,5,4,4,4,3,6,2,9,8,1,2,3,0,9,8,7,8,7,9,9,2,7,2,4,4,2,8,4,9,0,9,1,8,8,
              8,4,5,8,0,1,5,6,1,6,6,0,9,7,9,1,9,1,3,3,8,7,5,4,9,9,2,0,0,5,2,4,0,6,3,6,8,9,9,1,2,5,6,0,7,1,7,6,0,6,
              0,5,8,8,6,1,1,6,4,6,7,1,0,9,4,0,5,0,7,7,5,4,1,0,0,2,2,5,6,9,8,3,1,5,5,2,0,0,0,5,5,9,3,5,7,2,9,7,2,5,
              7,1,6,3,6,2,6,9,5,6,1,8,8,2,6,7,0,4,2,8,2,5,2,4,8,3,6,0,0,8,2,3,2,5,7,5,3,0,4,2,0,7,5,2,9,6,3,4,5,0]

1000-element Vector{Int64}:
 7
 3
 1
 6
 7
 1
 7
 6
 5
 3
 ⋮
 7
 5
 2
 9
 6
 3
 4
 5
 0

In [6]:
function ProductInVector(digits)
    
    #' @@name ProductInVector
    #'
    #' @@description
    #' 
    #' This function finds the largest product of cosecutive digits in vector all_digits.
    #' The product is determined by the amount of the input digits
    #'
    #' @@arg digits: A positive integer that determines how many cosecutive digits will be multiplied
    #'
    #' @@return The maximum product of cosecutive digits
    #' @@return The vector of all the consecutive digits that will be multiplied
    #' @@return The index of the first product in the vector all_digits
    
    max_product = 1
    index = 0
    vector = []
    
    for i in 1:length(all_digits)
        product = 1
        for j = 1:digits

            if ((i + j - 1) > 1000)
                break
            end

            product = product*all_digits[i + j - 1]
        end

        if product > max_product
            max_product = product
            index = i
        end
    end

    for k = 1:digits
        vector = push!(vector, all_digits[index + k - 1])
    end

    return max_product, vector, index
end

ProductInVector (generic function with 1 method)


let's now test our code with the given example

In [7]:
ProductInVector(4)

(5832, Any[9, 9, 8, 9], 616)

Therefore we get that the maximum product of four adjacent digits is $5832 = 9 \times 9 \times 8 \times 9$ and this is occurs at the index 616 in `all_digits`.

Now let's find the maximum product of thirteen adjacent digits 

In [8]:
ProductInVector(13)

(23514624000, Any[5, 5, 7, 6, 6, 8, 9, 6, 6, 4, 8, 9, 5], 198)

Therefore we get that the maximum product of thirteen adjacent digits is $23,514,624,000 = 5 \times 5 \times 7 \times 6 \times \times 6 \times 8 \times 9 \times 6 \times 6 \times 4 \times 8 \times 9 \times 5$ and this product occurs at index 198.

![](./images/image2.jpg){width=65%}
